Source/WebCore/platform/text/TextSpacing.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "TextSpacing.h"
+
+namespace WebCore {
+
+using namespace TextSpacing;
+
+bool TextAutospace::shouldApplySpacing(char32_t firstCharacter, char32_t secondCharacter) const
+{
+    auto firstCharacterClass = characterClass(firstCharacter);
+    auto secondCharacterClass = characterClass(secondCharacter);
+
+    if (hasIdeographAlpha()) {
+        if (firstCharacterClass == CharacterClass::Ideograph)
+            return secondCharacterClass == CharacterClass::NonIdeographLetter;
+        if (firstCharacterClass == CharacterClass::NonIdeographLetter && secondCharacterClass == CharacterClass::Ideograph)
+            return secondCharacterClass == CharacterClass::Ideograph;
```

Comment:
If you make the character classes each a separate bit (see BreakLines::classify & co), then you can reduce this to a single conditional check: `return (firstCharacterClass | secondCharacterClass == CharacterClass::NonIdeographLetter | CharacterClass::Ideograph)`.